Adds copy to clipboard custom service.

luixal opened the pull request:
## Description

Adds a custom service that adds an icon button on the indicator to copy to the clipboard the text in the `clipboard` field.

Fixes #500 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] I've read & comply with the [contributing guidelines](https://github.com/bastienwirtz/homer/blob/main/CONTRIBUTING.md)
- [x] I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers. 
- [x] I have made corresponding changes to the documentation (README.md).
- [x] I've checked my modifications for any breaking changes, especially in the `config.yml` file

luixal commented:
I guess this can also be made easily on the generic component. I can change it to that kind of implementation, but maybe it's better to keep things splitted apart.

Consider adding the hacktoberfest labels! :)

bastienwirtz commented on src/components/services/CopyToClipboard.vue:
```diff
@@ -0,0 +1,45 @@
+    copy() {
+      javascript:navigator.clipboard.writeText(this.item.clipboard)
+      this.animate = true;
+      setTimeout(
```
I think it shoud be possible to avoid the setTimeout by using CSS [`animation-delay`](https://developer.mozilla.org/en-US/docs/Web/CSS/animation-delay), that would be great.

luixal replied:
I implemented it this way to use the font-awesome animation class. I can  remove this by including a custom animation in CSS. It's related to the other change about making the animation shorter.

bastienwirtz commented on src/components/services/CopyToClipboard.vue:
```diff
@@ -0,0 +1,45 @@
+      this.animate = true;
+      setTimeout(
+        () => { this.animate = false },
+        1000
```
I not against the animation but I think it should be very fast, to it's discreet & subtle, you know what I mean ?

luixal replied:
As said before, I didn't control animation speed to avoid including a custom CSS animation (which I think it's better, just a fast "blink" or pulse) just because I though you would prefer it that way.

I agree to your comments so I'll include the custom animation in CSS :)

bastienwirtz commented on src/components/services/CopyToClipboard.vue:
```diff
@@ -0,0 +1,45 @@
+
+<style scoped lang="scss">
+.status {
+  color: black;
```
An alternative color for dark mode would be great!

bastienwirtz replied:
Also, maybe a litlle change of color on hover ? (has the entire card is clickable, it would confirm that you are about to activate on the action button)

luixal replied:
Yep, lessons learned from the proxmox service changes. I will look into bulma modifier classes because I'm not sure how to include a CSS selector that takes in count the `is-dark` class outside the component (remember, I'm not really good at frontend stuff...).

bastienwirtz reviewed (changes requested):
Thanks for the implementation @luixal. I think we can keep this as a separate service.
A few minor things, aside from that it looks good.

luixal commented:
Updated. As the animation is fast now, I removed the color alternation to green. Also, removed the timeout, using `animationend` event listener instead. Also included styles for light and dark themes.

Looks like this now:

![chrome-capture-2022-9-17](https://user-images.githubusercontent.com/615268/196122319-066d0232-bf09-4b2d-82f2-f229754a35af.gif)

this for dark mode:

![chrome-capture-2022-9-17 (1)](https://user-images.githubusercontent.com/615268/196122475-321c5527-f8e5-4a9a-9f4b-c1a21715e6c7.gif)

what do you think?

bastienwirtz commented on src/components/services/CopyToClipboard.vue:
```diff
@@ -0,0 +1,83 @@
+  },
+  methods: {
+    copy() {
+      javascript:navigator.clipboard.writeText(this.item.clipboard)
```
Is there a reason to use the `javascript:` prefix? I think this should be fine:  

```suggestion
      navigator.clipboard.writeText(this.item.clipboard)
```

luixal replied:
True, there's no need to use the `javascript:` prefix. I just took the code from the example in the issue. Removed.

bastienwirtz commented on src/components/services/CopyToClipboard.vue:
```diff
@@ -0,0 +1,83 @@
+  }),
+  mounted: function() {
+    this.$refs.copyIcon.addEventListener(
+      'animationend',
```
If you use `@animationend`, you should be able to remove this block

luixal replied:
True, no need for setting the listener manually.

bastienwirtz commented on src/components/services/CopyToClipboard.vue:
```diff
@@ -0,0 +1,83 @@
+<template>
+  <Generic :item="item">
+    <template #indicator>
+      <div class="status"><i ref="copyIcon" class="fa-regular fa-copy fa-xl" :class="{'scale': animate}" @click="copy()"></i></div>
```
I believe you can use `@animationend` instead of the `ref` + `mounted` function. 
```suggestion
      <div class="status"><i @animationend="animate=false" class="fa-regular fa-copy fa-xl" :class="{'scale': animate}" @click="copy()"></i></div>

```

luixal replied:
True. Removed.

bastienwirtz reviewed (commented):
Looks good to me :+1:

luixal commented:
Hi @bastienwirtz , were (or are) you able to merge this? As I accidentally removed my previous fork, I'm not sure how can this be handled.

luixal commented:
Will do that. I've also made a mess of the new fork, in fact, two of my PRs look just like the same (I swear I didn't mean to replicate it) so, once you've merged one of them (or discard, as you wish), close the other and I will clean all stuff up.

Meanwhile, I'll make another fork an recover branches with the patches, not sure if it will take the same PR or a different one.

Sorry for this mess...

luixal commented:
Well... it seems that if I put the same to the branch, github takes it at the same PR.

Changes done! :)

bastienwirtz commented:
@luixal hum I'm still unable to merge this request :cry: 
You may have to close and reopen the pull request. 

(If you want I can apply the patch myself but it won't appear as your contribution)

luixal commented:
I couldn't make the patch thing work, I copy&paste changes to my branch.

Feel free to apply the patch. Same for the other PR. I made a total mess with this xD

bastienwirtz commented:
@luixal Patch added to the main branch :+1:
